src/client_side_reply.cc

```diff
@@ -503,6 +503,10 @@ clientReplyContext::handleIMSReply(StoreIOBuffer result)
         http->logType.update(LOG_TCP_REFRESH_FAIL_ERR);
         debugs(88, 3, "origin replied with error " << status <<
                ", forwarding to client due to fail_on_validation_err");
+
+        if (collapsedRevalidation)
+            http->storeEntry()->clearPublicKeyScope();
+
         sendClientUpstreamResponse();
```

Comment:
> collapsed revalidation was not finalized when there was a negative response

... and is still not finalized in several other cases!

0. My minor PR commit 55cb158 covers all future sendClientUpstreamResponse() cases, leaving two existing areas exposed:

1. We probably need to handle sendClientOldEntry() cases. We probably do not want future collapsed revalidation clients to accidentally collapse on the entry that is being rejected by the code calling sendClientOldEntry() now. That future client should start its own revalidation attempt instead of collapsing on the stale entry created by an old attempt. To accomplish that, we need to mark the entry so that the already collapsed clients can still use it in their handleIMSReply() code while new clients cannot collapse on it. I believe releaseRequest(true) would do that. What do you think?

2. Consider cases where a collapsed revalidation did not reach handleIMSReply() at all. For example, imagine a client that created a collapsed revalidation entry and then disconnected or died. How to protect other clients from hitting that essentially stale entry forever, keeping bug 5051 alive? Unfortunately, I cannot think of a better solution than to add a collapsing timeout -- startCollapsingOn() would return false for collapsing-required entries that were around for longer than X seconds. Can you think of a better solution? If not, what X value would you recommend? We can hard-code X, for now, but if there is an existing configuraton directive that we can reuse for X, we should use that directive.

FWIW, bug 5051 should probably be described as "Squid does not check expiration of collapsed revalidation entries" rather than "Squid does not check expiration of negatively-cached collapsed-revalidation entries" -- negatively-cached collapsed-revalidation entries is just one (albeit common) manifestation of this bug. 

Our commit 1a210de essentially introduced a new class of publicly discoverable entries, but there is no code to check whether those entries have expired. For regular (i.e. not collapsable-on internal-revalidation) entries, such checks probably live in clientReplyContext::cacheHit(). Not all existing cacheHit() checks apply to (collapsed) revalidation, but some probably do.

Reply:
I really don't understand much about how collapsed revalidation is implemented.  I don't know what clearing a public key scope has to do with it.  I do see that clearPublicKeyScope() calls forcePublicKey() which calls clearPrivate() which sets shareableWhenPrivate=false, and that releaseRequest() calls setPrivateKey() which can set the same value.  releaseRequest() is called from abort(), so does that already handle the missing cases you are concerned about?

I don't understand why we wouldn't be seeing much more problems if all those cases you list are broken.  Even with the commit you already added to this PR, there's an extra call to  clearPublicKeyScope() when there is a 304 Not Modified response.  We get those all the time, so I can't imagine that collapsed revalidation is not getting properly cleared in that case.  And sendOldClient() calls restoreState(); does that end up also clearing collapsed revalidation from the storage?

Reply:
> releaseRequest() is called from abort(), so does that already handle the missing cases you are concerned about?

No, that only covers cases where the entry is abort()ed. I am concerned about completed entries.

> Even with the commit you already added to this PR, there's an extra call to clearPublicKeyScope() when there is a 304 Not Modified response. We get those all the time, so I can't imagine that collapsed revalidation is not getting properly cleared in that case.

Yes, that "extra" call did not fix any bugs in the current code. My polishing change avoids similar bugs in _future_ sendClientUpstreamResponse() callers. It is quite common for two calls that must go together to get separated during code refactoring, or for one of the calls to be forgotten. The latter is exactly what happened when we added the first clearPublicKeyScope() call. That is why it is a good idea to keep these two calls together.

> sendClientOldEntry() calls restoreState(); does that end up also clearing collapsed revalidation from the storage?

I was hoping that it does, and I did check it before posting my review.  AFAICT, it does not: restoreState() restores the current transaction state -- to use the old entry -- but simply unlocks the new entry. Due to old code design problems, it is difficult to be sure here, especially without looking at the logs, but I think the code will eventually end up in Store::Controller::handleIdleEntry() which does not seem to treat collapsed revalidation entries specially. If they are cachable, they may be left alone in the cache. I wonder whether we should add another clearPublicKeyScope() to that method because an idle entry cannot require collapsing (in a non-SMP environment)!


> I don't understand why we wouldn't be seeing much more problems if all those cases you list are broken.

That is a good question. There are actually not so many cases that remain allegedly buggy:

1. sendClientOldEntry() cases: There are probably only two relevant ones (the others do not deal with a cachable response). You can probably reproduce the first one -- "origin replied ... but with an older date header" --  using a custom script, but such cases are very rare in the wild. IIRC, that code was added primarily to make compliance tests happy. The second case would require a false failOnValidationError setting so it does not apply to your environment. 

2. A lack of handleIMSReply() call. To trigger this, one may have to kill the client-serving "job" after the negative response was cached but before that response was delivered to handleIMSReply(). I suspect it would be very difficult to time the client connection closure just right to accomplish this, so I am not very surprised that you do not see such cases. The "job" cleanup code may end up in Store::Controller::handleIdleEntry(). See my earlier paragraph about that method.

I wish we could clearPublicKeyScope() much earlier, when the response was received. It would solve all these problems before they start! Unfortunately, simply doing that would destroy the old being-revalidated entry -- there can be only one entry with the same public key (and default key scope).

I wonder whether we should move (part of) the logic that invalidates the old entry from handleIMSReply() to the code that caches the new entry. If we can make the right/informed decision about old entry preservation when the new entry is about to be cached, then it would be OK to invalidate the old entry early, of course. Unfortunately, I cannot volunteer to do the legwork for this non-trivial change right now.

If you would rather not do it either, then I can add the simpler changes (discussed in my review and this comment) and have you test them before this PR is merged. I will also add a TODO comment for the long-term fix described in the above paragraph.

How do you want to proceed?

Reply:
Let's go with the simpler changes and the TODO, and I will test them.

Reply:
Agreed. The ball is on my side.

Reply:
FYI: I started working on this earlier today. I hope to finish within 24 hours.

Reply:
Status update: I thought I was done adding the promised changes, but I am getting surprising test results now. I need more time to convince myself that these problems are _not_ related to this specific bug (and, hence, should be addressed separately). I hope to finish by Monday.

Reply:
The near-minimal changes discussed on this thread are now completed (as of branch commit cd35900).

Reply:
> I need more time to convince myself that these problems are not related to this specific bug (and, hence, should be addressed separately).

I plan to submit a small PR dedicated to those other problems. They are not related to stale collapsed revalidation responses that this PR is working on.

Reply:
> I plan to submit a small PR dedicated to those other problems

Done: #686 .